<a href="https://colab.research.google.com/github/pilaaaar/ANALISIS-NUMERICO/blob/main/RANDOM_FOREST.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

# 1. Cargar la base de datos
df = pd.read_csv('/content/DM.csv')

# 2. Eliminar las columnas no informativas
df = df.drop(columns=['EMG', 'Participante'])

# Mostrar las primeras 5 filas del DataFrame modificado
print("Primeras 5 filas del DataFrame después de eliminar columnas:")
display(df.head())

# Mostrar información general del DataFrame (tipos de datos, valores nulos)
print("\nInformación del DataFrame:")
df.info()


Primeras 5 filas del DataFrame después de eliminar columnas:


,Grupo,Minuto,C3,C4,CZ,F3,F4,F7,F8,FP1,...,O1,O2,P3,P4,PZ,ROG,T3,T4,T5,T6
0,0,1,1.457191,1.465362,1.389850,1.471585,1.383370,1.416681,1.335028,1.172173,...,1.419505,1.460171,1.435626,1.401177,1.405440,1.190718,1.486129,1.455536,1.457664,1.464154
1,0,10,1.442564,1.372664,1.361794,1.477391,1.366440,1.439025,1.269399,1.220473,...,1.417780,1.421081,1.410483,1.395753,1.393475,1.116700,1.474367,1.413287,1.477452,1.412524
2,0,11,1.445738,1.397893,1.363018,1.451693,1.372738,1.398308,1.285700,1.211166,...,1.444138,1.434578,1.425999,1.411785,1.403913,1.118753,1.472099,1.433305,1.484939,1.440299
3,0,12,1.381500,1.422462,1.443394,1.422085,1.413301,1.450779,1.404926,1.343141,...,1.427840,1.451817,1.403597,1.434047,1.428170,1.167026,1.428117,1.453502,1.373695,1.437340
4,0,13,1.443665,1.355041,1.360214,1.474152,1.365256,1.433039,1.281785,1.230395,...,1.415445,1.402281,1.391482,1.395755,1.400138,1.139237,1.464603,1.404249,1.453814,1.405438



Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 900 entries, 0 to 899
Data columns (total 23 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Grupo   900 non-null    int64  
 1   Minuto  900 non-null    int64  
 2   C3      900 non-null    float64
 3   C4      900 non-null    float64
 4   CZ      900 non-null    float64
 5   F3      900 non-null    float64
 6   F4      900 non-null    float64
 7   F7      900 non-null    float64
 8   F8      900 non-null    float64
 9   FP1     900 non-null    float64
 10  FP2     900 non-null    float64
 11  FZ      900 non-null    float64
 12  LOG     900 non-null    float64
 13  O1      900 non-null    float64
 14  O2      900 non-null    float64
 15  P3      900 non-null    float64
 16  P4      900 non-null    float64
 17  PZ      900 non-null    float64
 18  ROG     900 non-null    float64
 19  T3      900 non-null    float64
 20  T4      900 non-null    float64
 21  T5      900

Ahora que hemos cargado los datos y eliminado las columnas innecesarias, el siguiente paso es preparar los datos para el modelo Random Forest. Esto incluye separar las características (X) de la variable objetivo (y), y luego dividir los datos en conjuntos de entrenamiento y prueba.

### Explicación detallada del código y reporte de resultados

#### **Paso 1: Carga de Datos y Limpieza Inicial**

**Código (db68c844):**

```python
import pandas as pd

# 1. Cargar la base de datos
df = pd.read_csv('/content/DM.csv')

# 2. Eliminar las columnas no informativas
df = df.drop(columns=['EMG', 'Participante'])

# Mostrar las primeras 5 filas del DataFrame modificado
print("Primeras 5 filas del DataFrame después de eliminar columnas:")
display(df.head())

# Mostrar información general del DataFrame (tipos de datos, valores nulos)
print("\nInformación del DataFrame:")
df.info()
```

**Explicación:**

Este bloque de código realiza las siguientes acciones:

1.  **`import pandas as pd`**: Importa la librería `pandas`, fundamental para el manejo y análisis de datos en Python, asignándole el alias `pd` para facilitar su uso.
2.  **`df = pd.read_csv('/content/DM.csv')`**: Carga el archivo CSV llamado `DM.csv` (ubicado en `/content/DM.csv`) en un DataFrame de pandas, que es una estructura de datos tabular similar a una hoja de cálculo, y lo guarda en la variable `df`.
3.  **`df = df.drop(columns=['EMG', 'Participante'])`**: Elimina dos columnas, 'EMG' y 'Participante', del DataFrame. Esto se hace bajo la suposición de que estas columnas no aportan información relevante para el modelo o análisis posterior, o contienen datos que no son útiles en su formato actual.
4.  **`display(df.head())`**: Muestra las primeras 5 filas del DataFrame `df` después de la eliminación de columnas. Esto permite una rápida inspección visual de los datos y la confirmación de que las columnas se han eliminado correctamente.
5.  **`df.info()`**: Imprime un resumen conciso del DataFrame, incluyendo el número de entradas, el número total de columnas, los nombres de las columnas, el número de valores no nulos en cada columna, el tipo de datos (Dtype) de cada columna y el uso de memoria. Esta es una herramienta crucial para identificar posibles valores faltantes (Non-Null Count) y verificar los tipos de datos.

**Reporte de Resultados (de la celda db68c844):**

```text
Primeras 5 filas del DataFrame después de eliminar columnas:
```

| | Grupo | Minuto | C3 | C4 | CZ | F3 | F4 | F7 | F8 | FP1 | FP2 | FZ | LOG | O1 | O2 | P3 | P4 | PZ | ROG | T3 | T4 | T5 | T6 |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| 0 | 0 | 0 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 |
| 1 | 0 | 0 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 |
| 2 | 0 | 0 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 |
| 3 | 0 | 0 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 |
| 4 | 0 | 0 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 | 0.00 |

```text
Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 900 entries, 0 to 899
Data columns (total 23 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----
 0   Grupo   900 non-null    int64  
 1   Minuto  900 non-null    int64  
 2   C3      900 non-null    float64
 3   C4      900 non-null    float64
 4   CZ      900 non-null    float64
 5   F3      900 non-null    float64
 6   F4      900 non-null    float64
 7   F7      900 non-null    float64
 8   F8      900 non-null    float64
 9   FP1     900 non-null    float64
 10  FP2     900 non-null    float64
 11  FZ      900 non-null    float64
 12  LOG     900 non-null    float64
 13  O1      900 non-null    float64
 14  O2      900 non-null    float64
 15  P3      900 non-null    float64
 16  P4      900 non-null    float64
 17  PZ      900 non-null    float64
 18  ROG     900 non-null    float64
 19  T3      900 non-null    float64
 20  T4      900 non-null    float64
 21  T5      900 non-null    float64
 22  T6      900 non-null    float64
dtypes: float64(21), int64(2)
memory usage: 161.8 KB
```

Este paso confirma que los datos se han cargado correctamente, las columnas no deseadas han sido eliminadas y el DataFrame está libre de valores nulos, lo cual es un buen punto de partida para el modelado.

#### **Paso 2: Separación de Datos en Entrenamiento y Prueba**

**Código (e7077575):**

```python
from sklearn.model_selection import train_test_split

# 3. Separar características (X) y variable objetivo (y)
X = df.drop('Grupo', axis=1)  # Todas las columnas excepto 'Grupo' son características
y = df['Grupo']              # 'Grupo' es la variable objetivo

# 4. Dividir los datos en conjuntos de entrenamiento y prueba
# Usaremos un 80% para entrenamiento y un 20% para prueba
# random_state asegura que los resultados sean reproducibles
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

print("Dimensiones del conjunto de entrenamiento:")
print(f"X_train: {X_train.shape}")
print(f"y_train: {y_train.shape}")

print("\nDimensiones del conjunto de prueba:")
print(f"X_test: {X_test.shape}")
print(f"y_test: {y_test.shape}")

print("\nBalance de clases en y_train:")
print(y_train.value_counts(normalize=True))

print("\nBalance de clases en y_test:")
print(y_test.value_counts(normalize=True))
```

**Explicación:**

Este bloque de código se encarga de preparar los datos para el entrenamiento y evaluación del modelo de aprendizaje automático:

1.  **`from sklearn.model_selection import train_test_split`**: Importa la función `train_test_split` de la librería `scikit-learn`, que es una herramienta estándar para dividir conjuntos de datos en subconjuntos de entrenamiento y prueba.
2.  **`X = df.drop('Grupo', axis=1)`**: Crea el conjunto de características (variables independientes) `X` eliminando la columna 'Grupo' del DataFrame `df`. `axis=1` indica que se está eliminando una columna.
3.  **`y = df['Grupo']`**: Crea la variable objetivo (variable dependiente) `y` seleccionando únicamente la columna 'Grupo' del DataFrame `df'. Esta es la variable que el modelo intentará predecir.
4.  **`X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)`**: Divide los conjuntos `X` e `y` en subconjuntos de entrenamiento (`X_train`, `y_train`) y prueba (`X_test`, `y_test`).
    *   `test_size=0.2`: Asigna el 20% de los datos al conjunto de prueba y el 80% restante al de entrenamiento.
    *   `random_state=42`: Fija la semilla para la generación de números aleatorios, lo que asegura que la división de los datos sea la misma cada vez que se ejecuta el código, garantizando la reproducibilidad de los resultados.
    *   `stratify=y`: Este parámetro es crucial en problemas de clasificación con clases desbalanceadas. Asegura que la proporción de las clases en la variable objetivo (`y`) se mantenga la misma tanto en los conjuntos de entrenamiento como de prueba. Esto evita que un conjunto tenga una representación de clase significativamente diferente, lo que podría sesgar el entrenamiento o la evaluación del modelo.
5.  **`print` statements**: Imprimen las dimensiones (número de filas y columnas) de los conjuntos `X_train`, `y_train`, `X_test`, `y_test` para verificar que la división se haya realizado correctamente. También muestran el balance de clases en `y_train` y `y_test` para confirmar que `stratify=y` funcionó adecuadamente.

**Reporte de Resultados (de la celda e7077575):**

```text
Dimensiones del conjunto de entrenamiento:
X_train: (720, 22)
y_train: (720,)

Dimensiones del conjunto de prueba:
X_test: (180, 22)
y_test: (180,)

Balance de clases en y_train:
Grupo
0    0.555556
1    0.444444
Name: proportion, dtype: float64

Balance de clases en y_test:
Grupo
0    0.555556
1    0.444444
Name: proportion, dtype: float64
```

Estos resultados confirman que los datos se dividieron en proporciones 80/20 y que la estratificación mantuvo el mismo balance de clases en ambos conjuntos, lo cual es fundamental para entrenar y evaluar un modelo de clasificación de manera justa.

In [ ]:
from sklearn.model_selection import train_test_split

# 3. Separar características (X) y variable objetivo (y)
X = df.drop('Grupo', axis=1)  # Todas las columnas excepto 'Grupo' son características
y = df['Grupo']              # 'Grupo' es la variable objetivo

# 4. Dividir los datos en conjuntos de entrenamiento y prueba
# Usaremos un 80% para entrenamiento y un 20% para prueba
# random_state asegura que los resultados sean reproducibles
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

print("Dimensiones del conjunto de entrenamiento:")
print(f"X_train: {X_train.shape}")
print(f"y_train: {y_train.shape}")

print("\nDimensiones del conjunto de prueba:")
print(f"X_test: {X_test.shape}")
print(f"y_test: {y_test.shape}")

print("\nBalance de clases en y_train:")
print(y_train.value_counts(normalize=True))

print("\nBalance de clases en y_test:")
print(y_test.value_counts(normalize=True))


Dimensiones del conjunto de entrenamiento:
X_train: (720, 22)
y_train: (720,)

Dimensiones del conjunto de prueba:
X_test: (180, 22)
y_test: (180,)

Balance de clases en y_train:
Grupo
0    0.555556
1    0.444444
Name: proportion, dtype: float64

Balance de clases en y_test:
Grupo
0    0.555556
1    0.444444
Name: proportion, dtype: float64


Con los datos divididos en conjuntos de entrenamiento y prueba, el siguiente paso es entrenar el modelo Random Forest utilizando los datos de entrenamiento y luego evaluarlo con los datos de prueba. Esto nos permitirá ver qué tan bien generaliza el modelo a datos que no ha visto antes.

Con el modelo entrenado y evaluado, hemos completado el proceso de aplicación de Random Forest para la predicción de la columna 'Grupo'. El reporte de clasificación, la matriz de confusión y la precisión nos dan una buena idea del rendimiento del modelo.

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# 5. Inicializar y entrenar el modelo Random Forest
# Usaremos un random_state para reproducibilidad
model = RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)

print("Modelo Random Forest entrenado exitosamente.")

# 6. Realizar predicciones sobre el conjunto de prueba
y_pred = model.predict(X_test)

# 7. Evaluar el rendimiento del modelo
print("\nReporte de Clasificación:")
print(classification_report(y_test, y_pred))

print("\nMatriz de Confusión:")
print(confusion_matrix(y_test, y_pred))

print("\nPrecisión (Accuracy):", accuracy_score(y_test, y_pred))


Modelo Random Forest entrenado exitosamente.

Reporte de Clasificación:
              precision    recall  f1-score   support

           0       0.93      0.91      0.92       100
           1       0.89      0.91      0.90        80

    accuracy                           0.91       180
   macro avg       0.91      0.91      0.91       180
weighted avg       0.91      0.91      0.91       180


Matriz de Confusión:
[[91  9]
 [ 7 73]]

Precisión (Accuracy): 0.9111111111111111


In [ ]:
from io import StringIO

# Capturar el reporte de clasificación en una cadena de texto
report_str = classification_report(y_test, y_pred, output_dict=False)

# Leer la cadena de texto como un DataFrame de pandas
# Usamos StringIO para tratar la cadena como un archivo y pd.read_csv para parsearla
# skipinitialspace=True es útil si hay espacios después de los delimitadores
report_df = pd.read_csv(StringIO(report_str), sep='\s{2,}', engine='python', index_col=0)

# Guardar el DataFrame a un archivo CSV
report_df.to_csv('classification_report.csv')

print("Reporte de clasificación exportado a 'classification_report.csv'")

Reporte de clasificación exportado a 'classification_report.csv'


<>:9: SyntaxWarning: invalid escape sequence '\s'
<>:9: SyntaxWarning: invalid escape sequence '\s'
/tmp/ipykernel_12583/3626163004.py:9: SyntaxWarning: invalid escape sequence '\s'
  report_df = pd.read_csv(StringIO(report_str), sep='\s{2,}', engine='python', index_col=0)


El archivo `classification_report.csv` ahora está disponible en tu entorno de Colab. Puedes descargarlo desde el panel de archivos en el lateral izquierdo o con el siguiente código:

In [ ]:
from google.colab import files
files.download('classification_report.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

#### **Reporte de Clasificación del Modelo Random Forest**

El reporte de clasificación es una métrica clave para evaluar el rendimiento de un modelo de clasificación, proporcionando una visión detallada de cómo el modelo se comporta para cada clase individual, así como un rendimiento general. Basado en los resultados de la celda `9ad70043`:

```
Reporte de Clasificación:
              precision    recall  f1-score   support

           0       0.93      0.91      0.92       100
           1       0.89      0.91      0.90        80

    accuracy                           0.91       180
   macro avg       0.91      0.91      0.91       180
weighted avg       0.91      0.91      0.91       180


Matriz de Confusión:
[[91  9]
 [ 7 73]]

Precisión (Accuracy): 0.9111111111111111
```

Aquí está el desglose de lo que significan estos resultados:

*   **Clases (0 y 1)**: Representan los dos grupos o categorías que el modelo está tratando de predecir.

    *   **Clase 0:**
        *   **Precision (0.93):** De todas las veces que el modelo predijo la Clase 0, el 93% de esas predicciones fueron correctas. Esto indica que el modelo es muy bueno evitando falsos positivos para la Clase 0.
        *   **Recall (0.91):** El modelo identificó correctamente el 91% de todas las instancias reales de la Clase 0. Esto significa que el modelo es bueno detectando la mayoría de los casos verdaderos de la Clase 0.
        *   **F1-score (0.92):** Es la media armónica de la precisión y el recall. Un valor de 0.92 indica un buen equilibrio entre precisión y recall para la Clase 0.
        *   **Support (100):** Hubo 100 instancias reales de la Clase 0 en el conjunto de prueba.

    *   **Clase 1:**
        *   **Precision (0.89):** De todas las veces que el modelo predijo la Clase 1, el 89% de esas predicciones fueron correctas. Esto es ligeramente más bajo que para la Clase 0, pero sigue siendo un buen resultado.
        *   **Recall (0.91):** El modelo identificó correctamente el 91% de todas las instancias reales de la Clase 1. Similar a la Clase 0, el modelo es efectivo en detectar la mayoría de los casos verdaderos de la Clase 1.
        *   **F1-score (0.90):** Un valor de 0.90 indica un buen rendimiento general para la Clase 1, aunque también ligeramente inferior a la Clase 0.
        *   **Support (80):** Hubo 80 instancias reales de la Clase 1 en el conjunto de prueba.

*   **Accuracy (0.91):** La precisión general del modelo es del 91%. Esto significa que el 91% de todas las predicciones (tanto para la Clase 0 como para la Clase 1) fueron correctas.

*   **Macro Avg (Precision, Recall, F1-score: 0.91):** La media no ponderada de cada métrica para todas las clases. Es útil cuando se desea tratar a todas las clases por igual, independientemente de su número de instancias.

*   **Weighted Avg (Precision, Recall, F1-score: 0.91):** La media ponderada de cada métrica, donde cada métrica se pondera por el `support` (número de instancias) de cada clase. Esto es útil para tener una visión general del rendimiento del modelo que tiene en cuenta el desbalance de clases (si lo hay).

**Conclusión:**

El modelo Random Forest demuestra un **excelente rendimiento** general con una precisión del 91%. Las métricas de precisión, recall y F1-score son consistentemente altas para ambas clases, lo que indica que el modelo es robusto y generaliza bien. La **matriz de confusión** también refuerza esto:

*   **91** predicciones correctas para la Clase 0 (Verdaderos Negativos).
*   **73** predicciones correctas para la Clase 1 (Verdaderos Positivos).
*   **9** errores donde la Clase 0 fue predicha como Clase 1 (Falsos Positivos).
*   **7** errores donde la Clase 1 fue predicha como Clase 0 (Falsos Negativos).

Estos resultados sugieren que el modelo es capaz de distinguir eficazmente entre las dos clases de 'Grupo'.

#### **Explicación de la Matriz de Confusión**

La matriz de confusión es una tabla que se utiliza para describir el rendimiento de un modelo de clasificación en un conjunto de datos para el que se conocen los valores verdaderos. Nos permite visualizar el rendimiento del algoritmo, especialmente cuando las clases están desequilibradas. A partir de los resultados de la celda `9ad70043`, la matriz de confusión es la siguiente:

```
Matriz de Confusión:
[[91  9]
 [ 7 73]]
```

Esta matriz se interpreta de la siguiente manera:

|                   | **Predicción: Clase 0** | **Predicción: Clase 1** |
|-------------------|-------------------------|-------------------------|
| **Real: Clase 0** | 91 (Verdaderos Negativos) | 9 (Falsos Positivos)    |
| **Real: Clase 1** | 7 (Falsos Negativos)    | 73 (Verdaderos Positivos) |

Desglose de cada componente:

*   **Verdaderos Negativos (True Negatives - TN): 91**
    *   El modelo predijo correctamente que 91 instancias pertenecían a la **Clase 0**, y estas instancias realmente eran de la Clase 0.
    *   Estas son las predicciones correctas donde el modelo predijo la clase negativa (Clase 0).

*   **Falsos Positivos (False Positives - FP): 9**
    *   El modelo predijo incorrectamente que 9 instancias pertenecían a la **Clase 1**, pero en realidad eran de la Clase 0.
    *   También conocido como error de "Tipo I". El modelo dio una alarma cuando no debería haberlo hecho.

*   **Falsos Negativos (False Negatives - FN): 7**
    *   El modelo predijo incorrectamente que 7 instancias pertenecían a la **Clase 0**, pero en realidad eran de la Clase 1.
    *   También conocido como error de "Tipo II". El modelo no detectó algo que debería haber detectado.

*   **Verdaderos Positivos (True Positives - TP): 73**
    *   El modelo predijo correctamente que 73 instancias pertenecían a la **Clase 1**, y estas instancias realmente eran de la Clase 1.
    *   Estas son las predicciones correctas donde el modelo predijo la clase positiva (Clase 1).

**Interpretación:**

*   El modelo ha clasificado correctamente 91 + 73 = 164 de las 180 instancias totales en el conjunto de prueba, lo que resulta en una precisión (accuracy) de 164/180 = 0.9111 o 91.11%.
*   Los **Falsos Positivos (9)** y **Falsos Negativos (7)** son relativamente bajos, lo que indica que el modelo comete pocos errores de clasificación. Es ligeramente mejor al evitar clasificar incorrectamente la Clase 0 como Clase 1 (9 FP) que al clasificar incorrectamente la Clase 1 como Clase 0 (7 FN).

En resumen, la matriz de confusión muestra que el modelo Random Forest tiene un muy buen rendimiento en la distinción entre las dos clases, con un bajo número de errores de clasificación en ambos tipos.